x/protorev/keeper/rebalance.go

```diff
@@ -0,0 +1,160 @@
+	var optimalRoute gammtypes.SwapAmountInRoutes
+	maxProfitInputCoin := sdk.NewCoin(types.OsmosisDenomination, sdk.ZeroInt())
+	maxProfit := sdk.ZeroInt()
+
+	for _, route := range routes {
+		// Find the max profit for the route using the token out denom of the last pool in the route as the input token denom
+		inputCoin, profit, err := k.FindMaxProfitForRoute(ctx, route, route[2].TokenOutDenom)
+		if err != nil {
+			k.Logger(ctx).Error("Error finding max profit for route: ", err)
+			continue
+		}
+
+		// Filter out routes that don't have any profit
+		if profit.LTE(sdk.ZeroInt()) {
+			continue
+		}
+
+		// If arb doesn't start and end with uosmo, then we convert the profit to uosmo, and compare profits in terms of uosmo
+		if inputCoin.Denom != types.OsmosisDenomination {
+			uosmoProfit, err := k.ConvertProfits(ctx, inputCoin, profit)
+			if err != nil {
+				k.Logger(ctx).Error("Error converting profits: ", err)
+				continue
+			}
+			profit = uosmoProfit
+		}
+
+		// Select the optimal route King of the Hill style (route with the highest profit will be executed)
+		if profit.GT(maxProfit) {
+			optimalRoute = route
+			maxProfit = profit
+			maxProfitInputCoin = inputCoin
+		}
+	}
+
+	return maxProfitInputCoin, maxProfit, optimalRoute
+}
+
+// ConvertProfits converts the profit denom to uosmo to allow for a fair comparison of profits
+func (k Keeper) ConvertProfits(ctx sdk.Context, inputCoin sdk.Coin, profit sdk.Int) (sdk.Int, error) {
+	// Get highest liquidity pool ID for the input coin and uosmo
+	conversionPoolID, err := k.GetOsmoPool(ctx, inputCoin.Denom)
+	if err != nil {
+		return profit, err
+	}
+
+	// Get the pool
+	conversionPool, err := k.gammKeeper.GetPoolAndPoke(ctx, conversionPoolID)
+	if err != nil {
+		return profit, err
+	}
+
+	// Calculate the amount of uosmo that we can get if we swapped the
+	// profited amount of the orignal asset through the highest uosmo liquidity pool
+	conversionTokenOut, err := conversionPool.CalcOutAmtGivenIn(ctx, sdk.NewCoins(sdk.NewCoin(inputCoin.Denom, profit)), types.OsmosisDenomination, conversionPool.GetSwapFee(ctx))
+	if err != nil {
+		return profit, err
+	}
+
+	// return the profit denominated in uosmo
+	return conversionTokenOut.Amount, nil
+}
+
+// EstimateMultihopProfit estimates the profit for a given route
+// by estimating the amount out given the amount in for the first pool in the route
+// and then subtracting the amount in from the amount out to get the profit
+func (k Keeper) EstimateMultihopProfit(ctx sdk.Context, inputDenom string, amount sdk.Int, route gammtypes.SwapAmountInRoutes) (sdk.Coin, sdk.Int, error) {
+	tokenIn := sdk.NewCoin(inputDenom, amount)
+	amtOut, err := k.gammKeeper.MultihopEstimateOutGivenExactAmountIn(ctx, route, tokenIn)
+	if err != nil {
+		return sdk.NewCoin(types.OsmosisDenomination, sdk.ZeroInt()), sdk.ZeroInt(), err
```

Comment:
I think we can deduce this to sdk.Coin{} unless we want to specific osmo denom in error